docs/advanced_topics/customisation/custom_user_models.md

```diff
@@ -81,3 +81,49 @@ WAGTAIL_USER_EDIT_FORM = 'users.forms.CustomUserEditForm'
 WAGTAIL_USER_CREATION_FORM = 'users.forms.CustomUserCreationForm'
 WAGTAIL_USER_CUSTOM_FIELDS = ['country', 'status']
 ```
+
+```{versionadded} 6.2
+The ability to customize the `UserViewSet` was added. Instead of customizing the forms via the `WAGTAIL_USER_EDIT_FORM`, `WAGTAIL_USER_CREATION_FORM`, and `WAGTAIL_USER_CUSTOM_FIELDS` settings, we recommend customizing them via the viewset instead, as explained below. The aforementioned settings might be deprecated in a future release.
+```
```

Comment:
The currently documented way to customise the user forms (added in #2414) is a bit iffy. For one, I don't understand why we suggest specifying the custom fields in `WAGTAIL_USER_CUSTOM_FIELDS`, if the fields are already added declaratively in the custom forms.

With the example in https://docs.wagtail.org/en/stable/advanced_topics/customisation/custom_user_models.html, adding `WAGTAIL_USER_CUSTOM_FIELDS = ['country', 'status']` really has no effect, as the fields are already specified declaratively on the form class. The setting is used here,

https://github.com/wagtail/wagtail/blob/a09bba67cd58f519f3ae5bff32575e7ce9244031/wagtail/users/forms.py#L27-L33

which is then combined with the default fields in the form's `Meta.fields`:

https://github.com/wagtail/wagtail/blob/a09bba67cd58f519f3ae5bff32575e7ce9244031/wagtail/users/forms.py#L199-L203

However, [as per Django docs](https://docs.djangoproject.com/en/stable/topics/forms/modelforms/#overriding-the-default-fields),

> `ModelForm` is a regular `Form` which can automatically generate certain fields. The fields that are automatically generated depend on the content of the `Meta` class and on which fields have already been defined declaratively. Basically, `ModelForm` will only generate fields that are missing from the form, or in other words, fields that weren’t defined declaratively.

Reply:
I think you're right... the docs should have covered defining the form fields as part of the class _or_ listing them in WAGTAIL_USER_CUSTOM_FIELDS, but both is redundant.

I don't mind either way whether we formally deprecate these settings now, or leave that for some point in the future - however, we should definitely remove them from these instructions, because there's no sense in telling people to do it one way then saying "or alternatively you could do it the way we actually recommend". I suggest moving the `versionchanged` note to the bottom of this section and changing "might be deprecated in a future release" to "will be deprecated in a future release". ("Future release" gives us enough leeway that we don't need the extra qualifier...)

Reply:
Thank you. I reorganised the docs in 765fd0e88d699f77029aa7d3fef6fbc363b2d156. I think it'd be great if we can start the deprecation in the same release (so we don't have to bump it to 8.0 if 7.0 happens much sooner) – I've added the warnings in 3cc7d8ce99098fde3fd1c662152abade169ebd22. I also added another refactoring commit for the `SearchArea` in 9be664166ff0debf9984c2e3bc0c11aba7fb6a84, I hope that's OK. The `wagtail.users.wagtail_hooks` file is looking much more straightforward now 😄